### ISSUE TITLE
Add early stop functionality for NN training

### ISSUE THREAD

**Issue by drlebedev** [deeplearning4j/deeplearning4j#714]
It would be nice to have some kind of early stopping as a part of network config. 

Ideally, it would be good to have listener, that could detect that required accuracy is achieved. Than it is to easily implement your own early stop algorithms. It is not hard to add additional abstract listener type, like EarlyStopListener. Than it would be possible to extend it with custom implementation. 

I can imagine some implementations like ScoreEarlyStopListener provided as a part of dl4j.

Proposed ScoreEarlyStopListener API is the following:

```{java}
new ScoreEarlyStopListener(Double threshold, int numOfIteration)
```

where:
1. **Double threshold** - the value threshold value for score to start termination counter.  
2. **int numOfIteration** - if score is lower than threshold for numOfIteration times than terminate training

Listener approach allows to implement sophisticated termination algorithms based on test data as well as have stateful listeners.

**Comment by AlexDBlack** [comment]
Functionality I would like for this:
- Have a listener that evaluates network performance (usually a test set - different from training data) and records network parameters (to disk? database? binary? json? etc) whenever the test set score improves vs. best score so far
- Support for continual training on data until arbitrary termination criteria is met
  - Termination criteria including number of iterations, total time, etc. @drlebedev's accuracy criteria would then be just another possible termination criteria
  - See also http://page.mi.fu-berlin.de/prechelt/Biblio/stop_tricks1997.pdf for discussion on termination criteria
  - Would also be nice if we could do boolean ops on termination criteria: i.e., something like AND(numIterations(GREATER_THAN,100),OR(accuracyOnTestSet(0.9),testSetScore_current>2*testSetScore_minimum))

**Comment by AlexDBlack** [comment]
Early stopping functionality was added here: https://github.com/deeplearning4j/deeplearning4j/pull/955
